shiki/WordPress-Android
# Add item selection to Giphy Picker

![selection](https://user-images.githubusercontent.com/198826/49143945-63151e00-f2b9-11e8-9667-deea1be2cbf5.gif)

This moves the Giphy picker feature (wordpress-mobile/WordPress-Android#8627) forward. The changes feature:

- Selecting multiple items
- Updating of the "Preview" and "Add" labels in the bottom bar

## GiphyMediaViewModel

The `GiphyMediaViewModel` has been modified to include `isSelected` and `selectionNumber` `LiveData` properties. These properties are observed by `GiphyMediaViewHolder`. They are however changed by the `GiphyPickerViewModel` only. The `GiphyPickerViewModel` manages all the selected items and their continuous selection numbers.

To make sure that `GiphyMediaViewHolder` will never modify `GiphyMediaViewModel`, the `GiphyPickerViewModel` was converted to an `interface` and a mutable variant, `MutableGiphyMediaViewModel`, was added. The mutable variant is only accessed mainly by `GiphyPickerViewModel`.

## GiphyMediaViewHolder

This `ViewHolder` encapsulates the scaling and selection number animations when the `GiphyMediaViewModel.isSelected` and `selectionNumber` properties change. To allow observing of these `LiveData` properties within the `ViewHolder`, it's been changed to a `ViewHolder` that is also a `LifecycleOwner`.

The `LifecycleOwner` logic is contained in an abstract class named `LifecycleOwnerViewHolder`.

## Testing

1. Create a [Giphy API key](https://developers.giphy.com/docs/) and paste it in your `gradle.properties` as the value for `wp.giphy.api_key`
2. Go to the Media Browser
3. Click on the _Add_ button at the top right. Select the _Choose from Giphy_ option
4. In the Giphy picker, type what you want to search for and press Enter
5. When the results are shown, click on any of the items. You can click on more than one.

Here are a few UI behaviours that can be verified:

- The numbers that are shown on the selected items should be continuous. The numbers should be updated if you deselect one.
- When items are selected, the "Add" and "Preview" labels at the bottom should show the number of items that are selected. For example, "Add 3" and "Preview 3".
- If you hit the bottom of the list, the bottom bar (when shown) should not cover it. I suggest "donatello" as a search term since it does not have a lot of results.
- Rotating the device should keep the selected items
- Changing the search term (and pressing Enter) will clear the selected items

## Known Issues

- If you test this long enough, you will eventually hit API rate limits. This will cause empty blocks in the grid.
- Error handling for network errors will be added later
- Performing search results while typing will be added later
